### DISCUSSION TITLE
`flax.linen.initializers.ones` error, "TypeError: ones() missing 2 required positional arguments: 'key' and 'shape'"

### DISCUSSION THREAD

**Discussion by dashstander** [google/flax#2216] [Q&A] [answered]
Hello, I'm trying Flax for the first time after using Haiku for a while. I'm sure that this is probably a really stupid error, but I'm stuck.

Here is my attempt at implementing and then initializing an RMSNorm module:

```
import jax
import jax.numpy as jnp
import flax.linen as nn


class RMSNorm(nn.Module):

    @nn.compact
    def __call__(self, x):
        eps = 1e-8
        dim = x.shape[-1]
        scale = jnp.power(dim, -0.5)
        g = self.param('g', nn.initializers.ones(), (dim,))
        norm = jax.nn.normalize(x, axis=-1) * scale
        return x / jax.lax.clamp(eps, norm) * g

key = jax.random.PRNGKey(314159)
norm = RMSNorm()
text_x = jnp.ones((1, 512))
variables = jax.jit(norm.init)(key, text_x)
normalizer = norm.bind(variables)
```

but it errors on the `variables = ` line with the following error:

```
[<ipython-input-18-874a681eebf8>](https://localhost:8080/#) in __call__(self, x)
     11         dim = x.shape[-1]
     12         scale = jnp.power(dim, -0.5)
---> 13         g = self.param('g', nn.initializers.ones(), (dim,))
     14         norm = jax.nn.normalize(x, axis=-1) * scale
     15         return x / jax.lax.clamp(eps, norm) * g

TypeError: ones() missing 2 required positional arguments: 'key' and 'shape'
```

except, I thought that the `key` argument was provided by flax and the `shape` argument is there? At least, this is how the initializers are used in the documentation, e.g. [code block 21 on this page](https://flax.readthedocs.io/en/latest/notebooks/flax_basics.html).

So, again, I'm sure this is something really stupid I'm missing. But I appreciate your help and your time

**Comment by marcvanzee** [comment, marked as answer]
`Module.param` expects:
* a name
* a function `fn` that has signature `fn(rng, *args)`
* `*args` that will be passed to `fn`

In your case you are not passing a function, but the result of applying `nn.initializers.ones()`, which is not the correct way of calling that function. So you should replace it with the following:

```
g = self.param('g', nn.initializers.ones, (dim,))
```

There is also something wrong with the way you call `jax.lax.clamp`, you should provide a `max` as well. See the [documentation](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.clamp.html).

**Reply by dashstander** [reply]
Ahh, I had a feeling it was something silly like that. Thank you so much!
